profiler/hotapp/main.go

```diff
@@ -0,0 +1,105 @@
+// Copyright 2018 Google Inc. All rights reserved.
+// Use of this source code is governed by the Apache 2.0
+// license that can be found in the LICENSE file.
+
+// [START hotapp]
+
+// Sample hotapp is a synthetic application that exhibits different types of
+// profiling hotspots: CPU, heap allocations, thread contention.
+package main
+
+import (
+	"log"
+	"runtime"
+	"sync"
+	"time"
+
+	"cloud.google.com/go/profiler"
+)
+
+// There are several threads continuously fighting for this mutex.
+var mu sync.Mutex
+
+// Some allocated memory. Held in a global variable to protect it from GC.
+var mem [][]byte
+
+func sleepLocked(d time.Duration) {
+	mu.Lock()
+	time.Sleep(d)
+	mu.Unlock()
+}
+
+func contentionImpl(d time.Duration) {
+	for {
+		mu.Lock()
+		time.Sleep(d)
+		mu.Unlock()
+	}
+}
+
+func contention(d time.Duration) {
+	contentionImpl(d)
+}
+
+func waitImpl() {
+	select {}
+}
+
+// Waits forever.
+func wait() {
+	waitImpl()
```

Comment:
Can you explain why you have separate functions for wait/waitImpl; contention/contentionImpl; alloc/allocImpl?

Reply:
Done.